### ISSUE TITLE
Logging only errors and warnings

### ISSUE THREAD

**Issue by oomek** [bo3b/3Dmigoto#74]
I think it's a good idea to add a switch that allow for logging only errors and warnings. My log file sometimes grows to enormous sizes. It takes ages for it to load in the N++ and scrolling through it to find an errors is time consuming.

**Comment by DarkStarSword** [comment]
Sounds like a pretty good idea to me

**Comment by bo3b** [comment]
As a general rule, the default level of logging should already be setup this way.

The goal is to have normal logging (calls=1, debug=0) be only errors and significant events like CreateDevice, and exceptions.


If there is a lot of cruft in the log, it should be removed from default logging.  I think we've mostly have gotten rid of super common log spewers, but I'm sure there are others that are still too chatty.

In particular, stuff that runs per frame should never be logged, except in debug=1. Similarly keypress or other common events should not be logged.


What are examples of junk that you think should be trimmed?  We can switch anything too chatty from LogInfo to LogDebug.


BTW- this is one of those things that really bugs me about hand-rolled logging instead of using industry standard packages like Log4J, or Log4C.  They already do super great multi level logging, so if we are going to be putting in more effort to improve the logging, we ought to just switch to something that is already complete.

**Comment by oomek** [comment]
My suggestions of removals from the default logging are as follows:
1. repeating the whole ini section in the log by default is unneccessary
2. driver profile settings
3. HackerDevice messages
4. successfully reloaded shaders
5. resource copy errors should nor spawn on each frame
6. RecreateCompatibleTexture: Recreating cached resource flood

Also I think that most of people debugging shaders are interested in most recent bugs after pressing F10 so maybe 3DM should overwrite the log instead of appending it.

**Comment by bo3b** [comment]
OK, thanks for those.

For #1, #2, we definitely need to keep those.  When we have problems that only show up on certain machines and cannot debug things directly, I need to know what settings they actually use, not what they think they were using.  These should only be printed once.

Seems like the biggest problem you see is the F10 reload repeats stuff too much, and since you use it a lot while shader hacking it piles up.

**Comment by DarkStarSword** [comment]
1 and 2 we probably do want to keep - we specifically added them because we were sick of getting incomplete bug reports and they provide us a lot of info that users don't think to tell us so we can spot mistakes.

3 and 5 - ok

5 and 6 - agreed in principle, but those messages are indicating that 3DMigoto is repeatedly trying an operation, which has the potential to become expensive or fragment VRAM, so I'd rather fix the underlying issues instead of just hiding it:
  - If a resource or view creation fails, remember that and don't try again unless the description changes.
  - If a resource or view description changes, rather than destroying and recreating it keep a pool of resources and select the matching one from that pool (e.g. this happened in one of the WATCH_DOGS2 updates that meant the game sometimes used a full resolution buffer and sometimes used a downscaled buffer, so 3DMigoto would destroy and recreate its resources every time the game switched between them. This is the correct behaviour for 3DMigoto, but is not ideal. I don't recall if it is still the case in the most recent patch - Ubisoft did a lot of optimisations here)

7 - ok

**Comment by oomek** [comment]
Ok, I get your point about 1 and 2.
5 and 6 could have an sufix (repeated on each frame) instead of flooding the log
So what about for calls=1 remove 3,4,5,6,7
and for calls=2 all above with just errors logged with one line of messages prone to flooding?

I would like to know also what is your view on the keeping just the recent messages in the log.

**Comment by DarkStarSword** [comment]
Just came across another related case for seeing "RecreateCompatibleBuffer: Creating cached resource" repeated endlessly that I'm noting here because it's another case that produces annoying logs that might be tempting to silence, but should not be -in fact it indicates a different problem and silencing the logs would have hidden that.

This can be reproduced with something like:

    [Resource_HUD_Depth_UAV]
    type = Buffer
    format = R32_FLOAT
    array = 1
    
    [Resource_HUD_Depth_CB]
    type = Buffer
    format = R32G32B32A32_FLOAT
    array = 1
    
    [CustomShader_Calculate_HUD_Depth]
    post Resource_HUD_Depth_CB = copy Resource_HUD_Depth_UAV

This (surprisingly!) works, but specifying a format for a constant buffer is not correct because they don't have simple formats and this produces mismatching buffer descriptions causing 3DMigoto to endlessly recreate the buffer and produces annoying logs indicating there is a problem. The correct way to fix it is to remove the overrides from [Resource_HUD_Depth_CB] and allow 3DMigoto to automatically set the correct parameters (alternatively, byte_width=16 could have been used instead of format). Silencing the logs would have hidden this mistake, but it would have incurred 1. a hidden performance cost, 2. VRAM fragmentation and 3. loss of whatever data was in the buffer. The logs might be annoying, but they show us something is wrong so we can focus on fixing that.

We might be able to make 3DMigoto smarter for a case like this (use the format to calculate ByteWidth instead of StructureByteStride), but the whole point of the overrides is to allow us to stop 3DMigoto from trying to be smart, so... I'll think about it :-p

The recreate messages definitely need to be clearer though - at the very least they should indicate which resource is being recreated, and preferably give enough information to work out why they are being recreated.



As for truncating the log on reload... I'm not sure. I do sometimes need to refer to an earlier part of the log from before a reload, especially to compare errors or the like, but usually I'm probably only looking at the most recent section and I can probably afford to make a copy of the log for cases where I do need to compare two runs. I use 'tail -f' to monitor the logs in real time on another system, but it can handle the log being truncated (it just makes a note that happened and carries on). The times when we definitely want a complete log is when we ask someone to send it to us to debug something, but in that case it is unlikely they would be using the config reload mechanism, so that might be ok too.

**Comment by oomek** [comment]
In that case do what Kodi does. Alternate between d3d11_log.txt and d3d11_log.old.txt

**Comment by bo3b** [comment]
I definitely do not want to get into making more logs, and rolling over logs and that kind of thing, unless we switch to using Log4C.  All of that is already built in Log4C, us reinventing the wheel is a complete waste of time.  

Log4C would also allow us to create different log targets for different levels, so for example shader compilation errors could go to normal log, and also target a specific shader log.    

<br>
Also of note, are you using the -tail option in Notepad++?  Especially for doing F10 reloads, you can have that in an alternate window and see what happened directly.  (tail -f, View menu)  

It seems like we can easily trim the reloaded shaders listing from debug=0 log.  That listing of normal shaders being loaded, that does not have any errors is not super helpful. I don't think I've ever needed that part.

<br>
However, it seems to me that we are putting the cart before the horse here.  Oomek- please describe your use case and what problem you see, rather than trying to solve the problem in advance.   

We can very likely improve your use case to something that is not annoying, without wasting a bunch of time on stuff that doesn't matter.

**Comment by oomek** [comment]
There isn't much info on the Log4C explaining how it would help the log to become less cluttered, or at least I couldn't find any.
First you agree, then disagree, so I'm not going to push anything. I can do the same for logs what I did in my 3DM build before: allowing shader names prefixes, and reloading the #includes without "touching" all the files (I sort my files by date so I can see what have I modified recently, I wouldn't like to loose that)
I do not care if 3DM takes 1-2 second longer to reload shaders diring the shaderhacking. I would never release the mod with my build included.

Notepad++ is becoming very sluggish when the file is larger than 100kb and refuses to open files larger than 200-300 something. Tail -f won't help in that case and for per-frame floods it will make the situation even worse. It's not a big deal to press F5 and ctrl+end to see the recent log, but still I have to scroll through the information I'm not really interested to see.

I know all of that is against your principles, but in any case anyone else apart from you is interested I can publish the fork with my changes.

**Comment by bo3b** [comment]
No, that is not against my principles, but you are requesting that we break the log for our use, in order to make it better for you.  That's not going to happen, the log is our primary debugging tool for helping people with broken setups and games.

If we can support other use cases without breaking something we need, we are happy to do that.

<br> 
Again though, let's back up.  You still have not offered what the exact problem is.  Like most of these discussions, you have jumped directly to the expected solution, without detailing the actual problem.  Please let us figure out the solution that works for all of our users.    

It would help a lot if you'd simply attach a log that you feel is too chatty and causing you problems. Having something specific to discuss is better than generalities.  

<br>
Notepad++ is a fairly weak tool.  It's convenient, but has not aged well.  It cannot open files greater than 2G at all.  I've not had it be particularly sluggish on even 500MB files however.  If you are seeing it be slow, that suggests a RAM or paging problem.  

I don't understand your usage scenario, but an alternate way to watch the log would be to install the Git tools, and use the BASH shell that comes with. You get most of the interesting Unix tools, including things like tail -f and grep.  So it's easy enough to pipe a tail -f through grep looking for specific errors.    

 It also will then only dump out info that you are interested in, skipping everything else in the log. This type of approach is generally far superior than trying to manage logs.

**Comment by oomek** [comment]
If I knew it earlier that the log file was made solely for your personal use I wouldn't have bothered you at the first place. I wasn't asking for breaking anything, The topic was about adding an option to log only errors.

There is no problem with 3DM, it's just a convenience to have only errors, if they occur after pressing f10, on one page without doing miles with the scroll wheel. let's take for example this log:
---is there any other website like pastebin without a file limit?---
It takes 10 pages from the bottom before I reach the warnings section.

2GB ? I cannot open anything above 250ish MB, n++ tells me that te file is to large. I doubt it's a memory issue 8GB and a page file on a SSD should be enough for a text file I think. Will look up that git tools, and check if It can be of any use.

**Comment by oomek** [comment]
Would you please at least add a resource name to the message:
RecreateCompatibleTexture: Recreating cached resource
as I'm not entirely sure what I'm doing wrong.

**Comment by bo3b** [comment]
Sorry for the confusion, but no, the log is not strictly for our use.  It's just *primarily* for our use, so I don't want to change it in ways that make it less useful.  

There is also always the tension of creeping featuritis.  It's super easy and fun to add new features and capabilities, and little tweaks and special conditions and workarounds and on and on... and super hard to keep the code from becoming a total wreck because of all the special cases.  Right now our code is well on the way to spaghetti land.

<br>

Some background for this issue, and my recommendation.  (Caveat: I've used Log4J extensively, but Log4C may or may not be the best logging package for Win32.  There are at least 10 other candidates.)  


Adding a way to report only errors in a different path is bad, because it leads even further down the path of  a broken logging approach. No offense to Chiri, but people writing their own logger is how I know they are junior programmers.  The people who write Log4J/C have spent something like 3 people x 5 years on it.  There is no chance we can write something better in our spare time. And the big problem is opportunity cost.  Spending time on reinventing something lame like logging means we lose that time for something more valuable and specific to our mission.

This is why I'm bringing this up.  We already have two levels of logging, adding a third goes down that slippery path of wasted time.  Log4C already has 10 levels of logging, ability to roll over logs automatically, ability to direct different log levels to different receivers/logs, ability to have multiple output channels, like system level log or Visual Studio log, no \n grief, time and level stamps.  The only drawback to using it is the couple of hours it takes to learn how to use it, and the maybe 10 hours to switch over to it.

In your mind, you don't understand why this easy thing is trouble. I hate to be the hard-ass, but in my mind, this is a huge long-term deal, and in my judgment, if we are adding features to logging, we should go whole hog and fix it completely. And unlock some serious power instead of creeping one-off stuff.

<br>

So the summary of all that is I think we definitely should try to improve your use case.  But, we need to do it our way, not your way. 

Again, what exactly is the problem you see?  You specifically only want to see shader compilation errors upon an F10? Please do not specify the solution.

To add a log to this post, I think you can just drag/drop onto the text field.  Not sure what max size will be.

**Comment by bo3b** [comment]
> Would you please at least add a resource name to the message:
RecreateCompatibleTexture: Recreating cached resource
as I'm not entirely sure what I'm doing wrong.

That's a good idea.  We'll add that.  We can also probably trim the logs of Create*Shader, except in cases of error.  Those are not generally useful, except for letting us know how the game functions, and that we are successfully replacing shaders.

**Comment by bo3b** [comment]
As a start here, I did some research on the assorted logging libraries available today, and have trimmed the list to just a few candidates.  Anything that was too old, or clearly not maintained, was skipped.  

In order of recommendation:
1. g3Log: https://github.com/KjellKod/g3log
This matches our current formatting which would simplify adding this.  It supports all of the primary features I'd expect.  It has a primary advantage of specifically trying to handle exceptions and crashes and cleanly finishing the logs.  This would be valuable as our hand-baked version loses some trailing logs, even with unbuffered.  Second favorite logger on GitHub.
2. splog: https://github.com/gabime/spdlog
This has the interesting characteristic of using python style output formatting, which I'm sure would be helpful for DarkStarSword.  It is a bit simpler, and would be easier to integrate.  It would require reformatting all out current output lines.  Code syntax is cleaner and more clear than macro style.  Favorite logger on GitHub.
3. log4cpp: https://sourceforge.net/p/log4cpp/codegit/ci/master/tree/
This has the advantage of using the hierarchical style of Log4J.  If no one else has used this style or is familiar with it, that would not necessarily be an advantage. Setup is configurable for different use cases.  Usage is simple, with perhaps code that is not clear enough that it is logging as opposed to normal code.  No GitHub repo means it is probably obsolete.

Two are skipped in particular.  Pantheios is well respected and probably OK, but seems to no longer be maintained.  Boost.Logv2 is also well regarded, but is definitely overkill for our use, is hard to setup and use, and we don't really want to start bringing in Boost libraries.

I think that performance is not critical for our purposes, because for normal play use we can't afford to have any logging whatsoever.  Logging during development or hacking will already have other performance hits.

**Comment by bo3b** [comment]
OK, I still don't understand the problem or workflow issue.  Here is an example d3d11_log.txt from Alien Isolation with standard calls=1, debug=0 logging that I would expect to use while hunting and hacking shaders.  Hunting =1.

Running for several minutes shows that we correctly do not log anything during the primary game loop of RunningFrameActions (every Present call).  We also don't see any extra loads or creations of objects, although some games might very well do that.  There are other log spots that can be too chatty, but we need examples.

Doing F10 reload works and introducing an error into a shader gives the low boop sound and an error in the log as expected.  The error is back two pages of text, before the d3dx.ini settings repeat.  We can't remove the d3dx.ini repeat, because it's legitimate to change settings, then reload, and we need to know settings.  

We could possibly reorder the reload so that the shaders are reloaded *after* the d3dx.ini is reloaded, and that would print at the tail of the log.

<br>

By far the simplest way to see errors immediately is to just use Git Bash or Cygwin or Win10 linux shell, (preferably another computer over network share), and do *tail -f d3d11_log.txt | grep error*  That will print only the error lines which are the interesting part.  For warnings too, do *grep 'error\\|warning'*

For example:

    G:\SteamLibrary\steamapps\common\Alien Isolation\ShaderFixes\22d0cc9ef81dc3b2-vs_replace.txt(24,5-8): error X3004: undeclared identifier 'dang'

Example log:
[d3d11_log.txt](https://github.com/bo3b/3Dmigoto/files/1209817/d3d11_log.txt)

**Comment by DarkStarSword** [comment]
> We could possibly reorder the reload so that the shaders are reloaded after the d3dx.ini is reloaded, and that would print at the tail of the log.

I can't quite remember my reasoning at the moment, but I specifically did it in that order so that changes in the ini file would be loaded before processing the shaders to avoid a situation where the user would have to press F10 twice in a row to get everything to work. It might have been because back then we didn't have the hunting soft disabled state and editing the ini file was necessary to enable hunting - if we did it in the other order the shader reload would have been skipped altogether on the first F10 press because the ini hadn't been reloaded yet and so hunting was still hard disabled, so you would have to press F10 once to reload the ini and a second time to reload the shaders. If that's the only situation I was thinking of then we probably can switch them nowadays now that we have the soft disabled toggle since people use that over editing hunting in the d3dx.ini.

> Running for several minutes shows that we correctly do not log anything during the primary game loop of RunningFrameActions (every Present call). We also don't see any extra loads or creations of objects, although some games might very well do that. There are other log spots that can be too chatty, but we need examples.

The logs oomek mentioned as 5 and 6 are definitely way too chatty and sorting those out have been on my TODO list for a while, but like I said earlier I don't want to hide them by changing them to LogDebug - I want to make 3DMigoto more intelligent so that it doesn't endlessly retry something that will continue to fail to fix 5 and add a resource pool to fix 6, which will result in only seeing those logs once unless there is something else wrong in the users config like the other situation I described above.

Do any of the logging libraries you mentioned have the ability to rate limited specific messages like we can do in the Linux kernel?

**Comment by bo3b** [comment]
> > We could possibly reorder the reload so that the shaders are reloaded after the d3dx.ini is reloaded, and that would print at the tail of the log.

> I can't quite remember my reasoning at the moment, but I specifically did it in that order so that changes in the ini file would be loaded before processing the shaders to avoid a situation where the user would have to press F10 twice in a row to get everything to work. It might have been because back then we didn't have the hunting soft disabled state and editing the ini file was necessary to enable hunting - if we did it in the other order the shader reload would have been skipped altogether on the first F10 press because the ini hadn't been reloaded yet and so hunting was still hard disabled, so you would have to press F10 once to reload the ini and a second time to reload the shaders. If that's the only situation I was thinking of then we probably can switch them nowadays now that we have the soft disabled toggle since people use that over editing hunting in the d3dx.ini.

I haven't looked at this for a very long time.  The code is not well suited to changing the order in any case, the shader reload happens immediately, the .ini reload happens asynchronously after a flag is set.  We could move that around, but I'd be hesitant of breaking stuff.

More to the point, this is not necessary at all.  The keys used for shader reload and .ini reload are different, and can be specified differently in the d3dx.ini.  So for the scenario where you don't want or need the .ini reload, just comment out the:
*; reload_config = no_modifiers VK_F10*

**Comment by bo3b** [comment]
> The logs oomek mentioned as 5 and 6 are definitely way too chatty and sorting those out have been on my TODO list for a while, but like I said earlier I don't want to hide them by changing them to LogDebug - I want to make 3DMigoto more intelligent so that it doesn't endlessly retry something that will continue to fail to fix 5 and add a resource pool to fix 6, which will result in only seeing those logs once unless there is something else wrong in the users config like the other situation I described above.
> 
> Do any of the logging libraries you mentioned have the ability to rate limit specific messages like we can do in the Linux kernel?

Not that I can see.  Even Boost::Log does not appear to directly support rate limiting.  g3log can do this with a custom sink, or using the LOG_IF, if you know of specific locations in advance.  spdlog also has conditional logging. The *LOG_IF*, or *log->warn_if* could rate limit if hit too often.

However, is this something that would make sense as a different log level?  It seems like this behavior is closer to a Warn level, as opposed to an Error level, and that would allow us to keep logging by default at Warn, but people can reconfigure to only show Error as desired.

**Comment by oomek** [comment]
I'm from UK and have not much knowledge about consequences of stroking a bear, which I just did I suppose.
Anyway, here is my half gig log file
[](https://github.com/oomek/3Dmigoto/blob/master/Logs/d3d11_log4.rar?raw=true)

Reordering the log so the ini comes first is a good idea, if it doesn't break things of course.

**Comment by DarkStarSword** [comment]
>>    Would you please at least add a resource name to the message:
>>    RecreateCompatibleTexture: Recreating cached resource
>>    as I'm not entirely sure what I'm doing wrong.
>
>That's a good idea. We'll add that.

Just in case anyone is looking at adding that thinking 'this should be trivial', beware that it is not otherwise I would have just added it when I wrote that code. The resource name is not available in the function that creates those logs so it will need to be propagated into it *if there is a resource name we can use at all*. This may also be a cached resource associated with a copy operation that has no name we can use (maybe the section name it came from and line in the ini file would work instead), or a stereo2mono intermediate resource that also lacks a name. I'll try to do something about it this week, mental wellbeing permitting.

**Comment by DarkStarSword** [comment]
>  Anyway, here is my half gig log file

FWIW vim opens that log file in three seconds on my system - I have no idea how that compares with notepad++

Collating the most common lines with:

    sort d3d11_log4.txt | uniq -c | sort -n

And (for now) limiting our attention to anything that appeared at least 2000 times gives:

       2135 drawssr.hlsl(24,16-48): warning X3206: 'Load': implicit truncation of vector type
       2135 drawssr.hlsl(25,9-68): warning X3206: implicit truncation of vector type
       2448 ssr.hlsl(68,16-48): warning X3206: 'Load': implicit truncation of vector type
       2772     assembling replacement ASM code with shader model ds_5_0
       4354     compiling replacement HLSL code with shader model ps_5_0
       7904 --------------------------------------------- BEGIN ---------------------------------------------
       7904 ---------------------------------------------- END ----------------------------------------------
       8490     compiling replacement HLSL code with shader model vs_5_0
      12844     compile result for replacement HLSL shader: 0
      14652       Texture2D.MipLevels = -1
      14652       Texture2D.MostDetailedMip = 0
      14652   View Type = Shader Resource
      14725     Format = R24_UNORM_X8_TYPELESS (46)
      16285     ViewDimension = TEXTURE2D
      16285 Resource copy CreateCompatibleView failed: 80070057
     102114     Height = 495
     102114     Width = 410
     103608     BindFlags = 0x40
     153753     BindFlags = 0x20
     284535 RecreateCompatibleTexture: Recreating cached resource
     372598     Format = R10G10B10A2_UNORM (24)
     474256     Format = D24_UNORM_S8_UINT (45)
     572010     BindFlags = 0x28
     616563     Format = R24G8_TYPELESS (44)
     634488     BindFlags = 0x8
     658978 Resource copy error: Could not create/update destination resource
     659740 RecreateCompatibleTexture: Creating cached resource
     724203 Original resource was:
     724203 Resource copy RecreateCompatibleTexture failed: 0x80070057
    1362577     Height = 991
    1362577     Width = 821
    1463247     SampleDesc.Count = 1
    1464691     ArraySize = 1
    1464691     CPUAccessFlags = 0x0
    1464691     MipLevels = 1
    1464691     MiscFlags = 0x0
    1464691     SampleDesc.Quality = 0
    1464691     Usage = 0
    1464691   Resource Type = Texture2D

The biggest problem is definitely the repeated messages about failing to create an R24G8 texture with render target bind flags that is disallowed (though not documented as such) by DirectX. We do need to see that message because it tells us what is wrong, but we don't need to see it over and over - so I think the first priority should be on having 3DMigoto remember that the resource creation failed, and either don't try again, or at least don't print the log again until the resource description changes.

"RecreateCompatibleTexture: Creating cached resource" is part of the above issue, but "Recreating cached resource" is separate. This is what I indicated earlier I feel that the best solution is to add a resource pool that 3DMigoto can select one out of, so that instead of a flood of these, we will see one message for each distinct resource description in the pool.

I think both of those account for all the log messages that appeared over 14000 times. Below that number we start getting into shader compile warnings and errors.

**Comment by DarkStarSword** [comment]
> The recreate messages definitely need to be clearer though - at the very least they should indicate which resource is being recreated, and preferably give enough information to work out why they are being recreated.

> > > Would you please at least add a resource name to the message:
> > > RecreateCompatibleTexture: Recreating cached resource
> > > as I'm not entirely sure what I'm doing wrong.
> > 
> > That's a good idea. We'll add that.
> 
> Just in case anyone is looking at adding that thinking 'this should be trivial', beware that it is not otherwise I would have just added it when I wrote that code. The resource name is not available in the function that creates those logs so it will need to be propagated into it if there is a resource name we can use at all. This may also be a cached resource associated with a copy operation that has no name we can use (maybe the section name it came from and line in the ini file would work instead), or a stereo2mono intermediate resource that also lacks a name. I'll try to do something about it this week, mental wellbeing permitting.

This is now done and checked into master - I ended up going with the ini section name and line to identify copy operations in the log in every case (I wasn't keen on adding logic to switch between using that and the resource name just for logging, and the ini line will include the resource name anyway). stereo2mono intermediate resources are tagged "(intermediate)" to distinguish them from the destination resource cache. I also added identification for custom resource substantiations by the ini section they relate to (I did not log the ini line that triggered the substantiation, as it is largely irrelevant for these and for these cases it is the source resource which is substantiated, wheras in a copy operation it is the destination, which I felt would have been confusing).

    Substantiated custom Buffer [Resource_HUD_Depth_UAV]
    Creating cached buffer [Present] resource_hud_depth_cb = copy resource_hud_depth_uav
    Failed to substantiate custom Texture2D [ResourceBar]: 0x80070057
      Resource Type = Texture2D
        Width = 100
        Height = 100
        MipLevels = 0
        ArraySize = 1
        Format = R24G8_TYPELESS (44)
        SampleDesc.Count = 1
        SampleDesc.Quality = 0
        Usage = 0
        BindFlags = 0x20
        CPUAccessFlags = 0x0
        MiscFlags = 0x0
    Creating cached buffer [ShaderOverride_83bd33475a729417] resource_unityperdraw = vs-cb3
    Substantiated custom Buffer [Resource_Inverse_MVP]
    Creating cached buffer [ShaderOverride_83bd33475a729417] vs-cb10 = copy resource_inverse_mvp
    Creating cached buffer [ShaderOverride_352d8d3c52aa9229] ps-cb10 = copy resource_inverse_mvp
    Creating cached resource [ShaderOverride_93e3a6cda25c688d] resource_cameradepthtexture = ps-t0
    Creating cached buffer [ShaderOverride_93e3a6cda25c688d] resource_unitypercamera = ps-cb1
    Substantiated custom Texture2D [ResourceFoo]
    Substantiated custom TextureCube [ResourceBaz]
    Creating cached resource [customshader3dvision2sbs] ps-t100 = stereo2mono bb
    Creating cached resource [customshader3dvision2sbs] ps-t100 = stereo2mono bb (intermediate)

@oomek - you are using your own custom build of 3DMigoto for development right? If you don't need us to do a release urgently we might try to work on some of the other issues and bundle them together.

**Comment by oomek** [comment]
Thanks for your recent commits Ian. I like the current form, way more decsriptive. There is no rush with compiling a new public release if you've got other things in proggress. I will build it patched with my diffs.

**Comment by DarkStarSword** [comment]
@bo3b do you know an easy way to have HRESULTs decoded in our logs? Either a function we can call to return a string, or do any of the logging libraries you looked at support this? I'm kinda sick of googling these or trying to remember the command to look them up (mind you, in 99% of cases they end up being "invalid argument"... maybe I should just memories that hex code :-p )

**Comment by bo3b** [comment]
> @bo3b do you know an easy way to have HRESULTs decoded in our logs? Either a function we can call to return a string, or do any of the logging libraries you looked at support this? I'm kinda sick of googling these or trying to remember the command to look them up (mind you, in 99% of cases they end up being "invalid argument"... maybe I should just memorise that hex code :-p )

I've always just used the lookup tool in VS.  *Tools->Error lookup*, but I think that is also a cmd line tool.

We could pretty easily do it programmatically.  There is a COM error class, like: *LPCTSTR errMsg = err.ErrorMessage();*  We'd want to use the COM class instead of the normal FormatErrorMessage thing, because all the DX stuff is COM.  I think maybe FormatErrorMessage doesn't do COM errors, but not sure.

Not sure we want to drag in COM stuff, but also not sure we don't.  The smartptr part of COM makes things a lot more reliable, and takes away some of the problems with leaked objects or us having to manage all these gritty details.

Like most of this stuff in Win APIs, there's at least four ways to do everything, and I'm not enough of an expert to know which is best.

**Comment by DarkStarSword** [comment]
Yeah, and it looks like a lot of the options don't have any emphasis on code readability and they distract from what is important. I feel like I'm missing something here - HRESULTs have been around forever >_<

**Comment by bo3b** [comment]
Pretty good answer, with a few minutes searching.  https://stackoverflow.com/questions/455434/how-should-i-use-formatmessage-properly-in-c?noredirect=1&lq=1

_com_error seems to be the preferred choice, but you can't do the code above, as the string is released upon object destruction.  

Microsoft drives me crazy with their lack of consistency and API stability.  Always amazing to me how successful they've been with such awful tools.  Wish we could stereo on Linux.

**Comment by DarkStarSword** [comment]
Yuck:
"//automatic cleanup when error goes out of scope"
Nice in theory, but the errorText pointer can outlive the _com_error object and point to garbage memory.

**Comment by DarkStarSword** [comment]
... and I just found exactly that prayer that has been in 3DMigoto since day 1:

    static char *readStringParameter(wchar_t *val)
    {
            static char buf[MAX_PATH];
            wcstombs(buf, val, MAX_PATH);
            RightStripA(buf);
            char *start = buf; while (isspace(*start)) start++;
            return start;
    }

For starters we can drop the whitespace stripping there - the ini parsing will have already done that for us, but that's not a big deal. The problem is that we are returning a char* into a buffer that will have been allowed to go out of scope once that function returns. If the compiler inlines it (and in all likelyhood it will since the function has been declared static and is fairly small) it will probably be fine since the scope of buf will move up one level to the calling function, but if one day it decides not to inline it for whatever reason we have a pointer to already freed memory.

**Comment by DarkStarSword** [comment]
I've decided I will release these changes in 1.2.64 now - I've already made some potentially risky changes in that (the new ini parser), and adding a resource pool to fix some of these logging issues is a slightly larger risk - and I want a version between these changes that we can use for a bisection test in case something does get broken.

**Comment by bo3b** [comment]
That sounds like a good plan to me.  

If I get around to switching over to a new logger library, I'll make a branch to do that work, and will plan to bump to 1.3 version, because it will impact every file.  I'll coordinate with you all for a good time to switch over.  Changes this big makes all diffs subsequent diffs hard because it introduces so much noise.  Not able to get to it now, but I'll update this thread when I can.

**Comment by DarkStarSword** [comment]
I've pushed up a resource pool implementation to master, which should fix all the logging issues that appeared 14000 or more times in your log:
- "Recreating cached resource" flood gone (debug level will show "Switching cached resource" instead, but nothing is logged at information level)
- Now remembers which resource descriptions failed to create and will not retry them
- If a resource is successfully created but the description we get back from DirectX doesn't match, we will now still remember the original description we tried to use instead of endlessly recreating it (this wasn't the goal of the resource pool, but it still eliminates the log flood I noted in this comment so I'm ok with it: https://github.com/bo3b/3Dmigoto/issues/74#issuecomment-320443324 )

I'm going to let this change stew for a while before doing another release - I've tested it in a few games and it seems fine (and this does have a real benefit for WATCH_DOGS2), but there are a couple of rare situations where it could theoretically produce slightly different behaviour from before so I want to put it through it's paces first (I very much doubt we will hit these, but it is not impossible).

That's the last major logging issue that was impacting me. Looking at your log the next most common ones are all messages from the compiler... would I be correct in thinking that the reason you are seeing these so often is because you are running a custom 3DMigoto build that recompiles every shader on reload instead of just the modified ones? If so I don't really consider that a problem since it is not in the main 3DMigoto build and I think we can probably close this issue?
